### ISSUE TITLE
performance: vim.validate, vim.startswith, vim.endswith are slow

### ISSUE THREAD

**Issue by hoofcushion** [neovim/neovim#28921]
### Problem

I have identified a performance issue in the Neovim Lua API, specifically with the `vim.validate` function. The function is used to validate the types and values of function arguments, which is crucial for stability and correctness. However, it appears to be significantly slower than alternative approaches, which could impact the overall performance of Lua scripts and plugins in Neovim. And the plugin developers may need to create a no validation version of that function again.

### Conclusion
The performance of `vim.validate` is a critical concern for Lua script developers and users of Neovim. By implementing the proposed solution, we can maintain the robustness of our code while significantly improving performance. This will benefit the entire Neovim community by enabling faster and more efficient Lua scripting.

### Steps to reproduce

1. Run the attached Lua script.
2. Observe the execution times for each function.

### Script for Benchmarking
Below is the Lua script used for benchmarking the performance of different validation approaches:

```lua
---@diagnostic disable
local function test_speed(title, t, fn)
  local start_time = os.clock()
  local i = 0
  while i < t do
    fn()
    i = i + 1
  end
  local end_time = os.clock()
  local time_taken = end_time - start_time
  print(title .. ": " .. t .. " times took " .. time_taken .. " seconds")
end

local function empty() end

function normal_one(s, suffix)
  vim.validate({ s = { s, 's' }, suffix = { suffix, 's' } })
  return #suffix == 0 or s:sub(-#suffix) == suffix
end

local function no_validation(s, suffix)
  return #suffix == 0 or s:sub(-#suffix) == suffix
end

local function only_create_table(s, suffix)
  empty({ s = { s, "s" }, suffix = { suffix, "s" } })
end

local function pre_created_table(s, suffix)
  return #suffix == 0 or s:sub(-#suffix) == suffix
end

local function _validate(fn, vali, length)
  return function(...)
    for i = 1, length do
      vali[i][1] = select(i, ...)
    end
    vim.validate(vali)
    fn(...)
  end
end

pre_created_table = _validate(pre_created_table, {{ nil, "s" }, { nil, "s" }}, 2)

local t = 10000000
test_speed("vim.endswith", t, function() normal_one("hello world", "world") end)
test_speed("no validation", t, function() no_validation("hello world", "world") end)
test_speed("only create table", t, function() only_create_table("hello world", "world") end)
test_speed("pre-created table", t, function() pre_created_table("hello world", "world") end)
```

### Expected behavior

The `vim.validate` function should perform validation without causing a significant performance penalty.

### Actual Results
The attached script benchmarks four different validation approaches:

```
➜  nvim git:(master) ✗ nvim --headless -l test.lua                                               
vim.endswith: 10000000 times took 4.43177 seconds
no validation: 10000000 times took 0.012763000000001 seconds
only create table: 10000000 times took 1.182221 seconds
pre-created table: 10000000 times took 2.674695 seconds
```

The results show that using `vim.validate` is substantially slower compared to not validating at all or using a pre-created table for validation. In this case, in function `vim.endswith`, to create validate table, it takes 100x of times of the running of the function itself, and to validate the input, it takes 180x of times.

### Proposed Solution
To improve the performance of the `vim.validate` function, I propose the following solution:

1. **Reuse Validation Tables**: Instead of creating a new table for each validation, we can pre-create a table and reuse it for multiple validations. This reduces the overhead of table creation and garbage collection.

2. **Optimize `vim.validate`**: The current implementation of `vim.validate` could be optimized to reduce the time complexity. This might involve better use of Lua's native type checking and avoiding unnecessary operations.

3. **Provide a Configurable Validation Setting**: Allow users to configure the level of validation required. For performance-critical applications, or the values are trustful, a no validation version function could be used for speed. 

4. Do solution 1. slowly and eventually completely replace `vim.validation`. This optimization solution is stable and infallible.

5. E.g., in the following code, if something like `--no-validation` flag was set, _validate will not execute, everything is unsafe by default.

```lua
local function _validate(fn, vali, length)
 if ... then -- A true condition will skip the validation step.
  return fn 
 end
  return function(...)
    for i = 1, length do
      vali[i][1] = select(i, ...)
    end
    vim.validate(vali)
    fn(...)
  end
end
```

6. Use `debug.getupvalue` to get the no validation version. Very useful when a trusty code block needs a performance improve. AKA `unsafe`.

```lua
local function _validate(func,vali)
 return function(...)
  local e=select("#",...)
  for i=1,e do
   vali[i][1]=select(i,...)
  end
  vim.validate(vali)
  func(...)
 end
end
local function get_safe(fn)
 return select(2,debug.getupvalue(fn,2))
end
local std={}
function std.print(str,str,str)
 print(str,str,str)
end
std.print=_validate(std.print,{{nil,"string"},{nil,"string"},{nil,"string"}})
local unsafe_print=get_safe(std.print)
unsafe_print(1,2,3) -- no error
std.print(1,2,3) -- error: expect string
```

7. Add a max validation count parameter to `vim.validate`, so it doesn't need to declare the parameter amount at start.
```lua
local function _validate(func,vali)
 return function(...)
  local e=select("#",...)
  for i=1,e do
   vali[i][1]=select(i,...)
  end
  vim.validate(vali,e) -- there
  func(...)
 end
end
```

### Additional Notes
- The results may vary depending on the system and Neovim configuration.
- Further profiling and optimization may be required to fine-tune the performance improvements.

### Neovim version (nvim -v)

NVIM v0.11.0-dev-49+g10601ac5f Build type: RelWithDebInfo LuaJIT 2.1.1713484068

### Vim (not Nvim) behaves the same?

no

### Operating system/version

Arch Linux x86_64, Linux 6.9.1-zen1-1-zen

### Terminal name/version

Alacritty

### $TERM environment variable

xterm-256color

### Installation

bob use nightly

**Comment by hoofcushion** [comment]
I am sorry, I added a wrong label for this issue, I don't know how to change it. Sorry.

**Comment by dundargoc** [comment]
`vim.validate` got a huge performance boost last week in https://github.com/neovim/neovim/pull/28758, are you using the a version that is after that got merged?

**Comment by hoofcushion** [comment]
Yes, I use nightly. I think it has nothing to do with that. It slows, mainly because of the recreation of validation table.

**Comment by hoofcushion** [comment]
I tested, use a pre-created validation table will boost a lot.

**Comment by dundargoc** [comment]
> Yes, I use nightly.

Yes, but which commit/version?

**Comment by hoofcushion** [comment]
> NVIM v0.11.0-dev-49+g10601ac5f Build type: RelWithDebInfo LuaJIT 2.1.1713484068

**Comment by dundargoc** [comment]
Right, I missed the version. Yes, that is after that PR.

**Comment by hoofcushion** [comment]
I am currently in the GMT+8 time zone. I need to get some rest now, so I won't be able to respond immediately. I'll check messages again when I wake up. Bye.

**Comment by hoofcushion** [comment]
I have had another benchmark of `vim.validate`, and ironically, `vim.endswith`, which uses `string.sub` to test the trailer part of a string is equal to the input, and it's considered to be faster than `string.find`. But it's much slower than the `string.find`, which makes `vim.endswith` to be meaningless.
```lua
---@diagnostic disable
local function test_speed(title,t,fn)
 local start_time=os.clock()
 local i=0
 while i<t do
  fn()
  i=i+1
 end
 local end_time=os.clock()
 local time_taken=end_time-start_time
 print(title..": "..t.." times took "..time_taken.." seconds")
end
local t=10^6
test_speed("string.find",t,function() string.find("xxxxxxxxx","x$") end)
test_speed("vim.endswith",t,function() vim.endswith("xxxxxxxxx","x") end)
```
The outputs:
```
string.find: 10000000 times took 1.413474 seconds
vim.endswith: 10000000 times took 4.41524 seconds
```

**Comment by hoofcushion** [comment]
Here are a few ideas I've been considering that could be implemented in the newer Neovim Lua standard library to enhance performance:

- **`purify(fn, valilist)`**
  - This function could add a parameter filter to a given function.
  - It uses `vararg` to iterate every argument and apply check for validation.
  - The `valilist` is always reused for performance.
  - Alternatively, inline validation logic could be implemented for speed, as it might be more efficient than a function call.

- **`add_unsafe(fn)`**
  - This function would remove the filter added by `purify`.
  - It could use `debug.getupvalue` to get the original function that was pasted in `purify`.
  - This would allow for a performance boost in scenarios where safe.
  - Another name considered is `depurify`.

- **`flat_validate(opt)`**
  - This is similar to `vim.validate`, but its input is a flat list.
  - It uses a loop like `for i=1, #opts, 3 do ... end` to get `val`, `types`, and `optional` to validate the given validation table.

- **`Remove optional in validate(opt)`**
  - For values that are not optional, an additional nil check for regular values is redundant.
  - Using "nil" as a type check could be more efficient.
  - Also work for flat_validate too.

- **`Parse the given type list to a type set`**
  - For example, in `purify(func, {1, {"number", "string"}})`, the given validation table could be parsed into a hash map like `{["number"] = true, ["string"] = true}` first.
  - This would greatly increase the type check speed for a union type.

**Comment by justinmk** [comment]
> `vim.endswith`, which uses `string.sub` to test the trailer part of a string is equal to the `suffix`, and it's considered to be faster than `string.find`. But it's much slower than the `string.find`, which makes `vim.endswith` to be meaningless,

Agreed. Also `vim.startswith`. Those need to be fixed, else they are useless.

**Comment by gpanders** [comment]
Aside from any performance issues, we can limit `vim.validate` to top-level/"public" functions. Imo there's no reason to use `vim.validate` over `assert` for internal/private/`local` functions. `assert` will always be faster.

**Comment by gpanders** [comment]
Considering this resolved by https://github.com/neovim/neovim/pull/28977. We can convert many existing instances of `vim.validate` to the new form introduced in that PR. The old form is still slow, but should only need to be used in rare cases.
